pkg/core/native_gas_test.go

```diff
@@ -52,3 +59,95 @@ func TestGAS_Roundtrip(t *testing.T) {
+	}
+	// deposit GAS for `signer` with lock until the next block
+	depositAmount := 100_0000 + (2+int64(nKeys))*transaction.NotaryServiceFeePerKey // sysfee + netfee of the next transaction
+	transferTx := transferTokenFromMultisigAccount(t, chain, notaryHash, gasHash, depositAmount, signer, int64(chain.BlockHeight()+1))
+	res, err := chain.GetAppExecResults(transferTx.Hash(), trigger.Application)
+	require.NoError(t, err)
+	require.Equal(t, vm.HaltState, res[0].VMState)
+	require.Equal(t, 0, len(res[0].Stack))
+
+	// save initial validators balance
+	balances := make(map[int]int64, testchain.ValidatorsCount)
+	for i := 0; i < testchain.ValidatorsCount; i++ {
+		balances[i] = chain.GetUtilityTokenBalance(testchain.PrivateKeyByID(i).GetScriptHash()).Int64()
+	}
+	ic := interop.NewContext(trigger.Application, chain, chain.dao, nil, nil, nil, nil, chain.log)
+	tsInitial := chain.contracts.GAS.TotalSupply(ic, nil).Value().(*big.Int).Int64()
+
+	// send transaction with Notary contract as a sender
+	tx := chain.newTestTx(util.Uint160{}, []byte{byte(opcode.PUSH1)})
+	tx.Attributes = append(tx.Attributes, transaction.Attribute{Type: transaction.NotaryAssistedT, Value: &transaction.NotaryAssisted{NKeys: uint8(nKeys)}})
+	tx.NetworkFee = (2 + int64(nKeys)) * transaction.NotaryServiceFeePerKey
+	tx.Signers = []transaction.Signer{
+		{
+			Account: notaryHash,
+			Scopes:  transaction.None,
+		},
+		{
+			Account: signer,
+			Scopes:  transaction.None,
+		},
+	}
+	tx.Scripts = []transaction.Witness{
+		{
+			InvocationScript: append([]byte{byte(opcode.PUSHDATA1), 64}, notaryNodes[0].SignHashable(uint32(testchain.Network()), tx)...),
+		},
+		{
+			InvocationScript:   testchain.Sign(tx),
+			VerificationScript: testchain.MultisigVerificationScript(),
+		},
+	}
+	b := chain.newBlock(tx)
+	require.NoError(t, chain.AddBlock(b))
+	checkBalanceOf(t, chain, notaryHash, int(depositAmount-tx.SystemFee-tx.NetworkFee))
+	singleReward := transaction.NotaryServiceFeePerKey * (nKeys + 1) / nNotaries
+	for _, notaryNode := range notaryNodesPublicKeys {
+		checkBalanceOf(t, chain, notaryNode.GetScriptHash(), singleReward)
+	}
+	for i := 0; i < testchain.ValidatorsCount; i++ {
+		newBalance := chain.GetUtilityTokenBalance(testchain.PrivateKeyByID(i).GetScriptHash()).Int64()
+		expectedBalance := balances[i]
+		if i == int(b.Index)%testchain.CommitteeSize() {
+			// committee reward
+			expectedBalance += 5000_0000
+		}
+		if testchain.IDToOrder(i) == int(b.PrimaryIndex) {
+			// primary reward
+			expectedBalance += tx.NetworkFee - int64(singleReward*nNotaries)
+		}
+		assert.Equal(t, expectedBalance, newBalance, i)
+	}
+	tsUpdated := chain.contracts.GAS.TotalSupply(ic, nil).Value().(*big.Int).Int64()
+	tsExpected := tsInitial +
+		5000_0000 - // committee reward
+		(tx.NetworkFee + tx.SystemFee) + // burn tx fees
+		(tx.NetworkFee - int64(singleReward*nNotaries)) + // primary node reward
+		int64(singleReward*nNotaries) // notaries reward
```

Comment:
Why not just `tsInitial + 5000_0000 - tx.SystemFee`? Network fee is redistributed completely. Before the patch we had excessive `singleReward*nNotaries` generated.

Reply:
This long expression has the same value as the one you provided. I thought that we need to explain clearly the meaning of each part, but it seems to be redundant. Fixed.

Reply:
To me it's a bit easier to understand, total supply changes because of mints and burns. Network fee is just being redistributed, it's out of scope.